File: dart/lib/src/nft/retrieve_nft_owner_v2.dart
import 'package:solana/solana.dart' hide RpcClient;
import '../rpc/rpc_client.dart';
import 'get_domain_mint.dart';

/// V2 function to retrieve the owner of a tokenized domain name
///
/// This function mirrors js/src/nft/retrieveNftOwnerV2.ts
///
/// [rpc] - The RPC client for Solana blockchain communication
/// [nameAccount] - The key of the domain name
///
/// Returns the owner public key, or null if not tokenized
Future<Ed25519HDPublicKey?> retrieveNftOwnerV2(
  RpcClient rpc,
  Ed25519HDPublicKey nameAccount,
) async {
  try {
    final mint = await getDomainMint(nameAccount);

    // Get the largest token accounts for this mint
    final largestAccounts = await rpc.getTokenLargestAccounts(mint.toBase58());
    if (largestAccounts.isEmpty) {
      return null;
    }

    // Get account info for the largest account
    final largestAccountInfo = await rpc.fetchEncodedAccount(
      largestAccounts[0].address,
    );

    if (!largestAccountInfo.exists) {
      return null;
    }

    // Parse token account data
    // Token account layout: mint(32) + owner(32) + amount(8) + ...
    final data = largestAccountInfo.data;
    if (data.length >= 72) {
      // Ensure we have enough data for amount
      final ownerBytes = data.sublist(32, 64);

      // Check amount (8 bytes, little endian) - should be 1 for NFT
      final amountBytes = data.sublist(64, 72);
      var amount = 0;
      for (var i = 0; i < 8; i++) {
        amount += amountBytes[i] << (i * 8);
      }

      if (amount == 1) {
        return Ed25519HDPublicKey(ownerBytes);
      }
    }

    return null;
  } on Exception {
    // Handle RPC errors (mint does not exist, etc.)
    return null;
  }
}

File: js/src/nft/retrieveNftOwnerV2.ts
import { AccountLayout } from "@solana/spl-token";
import { Connection, PublicKey } from "@solana/web3.js";

import { getDomainMint } from "./getDomainMint";

/**
 * Retrieves the owner of a tokenized name using the mint's largest token account.
 *
 * Returns `null` when the mint or a one-token holder cannot be found.
 *
 * @param connection Solana RPC connection used to query token accounts.
 * @param nameAccount Tokenized SNS name account address.
 * @returns The owner public key, or `null` when no one-token holder exists.
 *
 * @example
 * ```ts
 * const owner = await retrieveNftOwnerV2(connection, nameAccount);
 * ```
 */
export const retrieveNftOwnerV2 = async (
  connection: Connection,
  nameAccount: PublicKey,
) => {
  try {
    const mint = getDomainMint(nameAccount);

    const largestAccounts = await connection.getTokenLargestAccounts(mint);
    if (largestAccounts.value.length === 0) {
      return null;
    }

    const largestAccountInfo = await connection.getAccountInfo(
      largestAccounts.value[0].address,
    );

    if (!largestAccountInfo) {
      return null;
    }

    const decoded = AccountLayout.decode(largestAccountInfo.data);
    if (decoded.amount.toString() === "1") {
      return decoded.owner;
    }
    return null;
  } catch (err) {
    if (
      typeof err === "object" &&
      err !== null &&
      "code" in err &&
      err.code === -32602
    ) {
      // Mint does not exist
      return null;
    }

    throw err;
  }
};
